### PR TITLE
Require Audio Easing support on RialtoMSEAudioSink for Netflix (audio-fade and fade-volume)

### PR DESCRIPTION
**Pull request by skyshaha1**
Summary: Require Audio Easing support on RialtoMSEAudioSink for Netflix - audio-fade and fade-volume
Type: Feature
Test Plan: UT
Jira: RIALTO-593

### CODE REVIEW

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 1]
```diff
@@ -377,6 +385,18 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
...
     }
+    case PROP_FADE_VOLUME:
+    {
+        if (!client)
+        {
+            g_value_set_uint(value, priv->fadeVolume);
+            return;
+        }
+
+        double fadeVolume{kDefaultFadeVolume};
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:396]: This needs to call getVolume

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 2]
```diff
@@ -521,6 +541,32 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
+        uint32_t duration;
+        int easeTypeInt;
+
+        if (sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt) != 3)
+        {
+            GST_ERROR_OBJECT(object, "Failed to parse audio fade string");
+            return;
+        }
+
+        firebolt::rialto::EaseType easeType = static_cast<firebolt::rialto::EaseType>(easeTypeInt);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:557]: Just a reminder, remove cast

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 3]
```diff
@@ -521,6 +541,32 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
         }
         break;
     }
+    case PROP_AUDIO_FADE:
+    {
+        const gchar *audioFadeStr = g_value_get_string(value);
+        double volume;
+        uint32_t duration;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:548]: Init these to the defaults

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 4]
```diff
@@ -521,6 +541,32 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
+        int easeTypeInt;
+
+        if (sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt) != 3)
+        {
+            GST_ERROR_OBJECT(object, "Failed to parse audio fade string");
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:553]: Maybe a warning and continue?

File: source/RialtoGStreamerMSEAudioSinkPrivate.h [hunk 5]
```diff
@@ -38,6 +47,10 @@ struct _RialtoMSEAudioSinkPrivate
     std::atomic_bool isSyncOffQueued = false;
     std::atomic<int32_t> streamSyncMode = kDefaultStreamSyncMode;
     std::atomic_bool isStreamSyncModeQueued = false;
+    std::atomic<const gchar*> audioFade = kDefaultAudioFade;
+    AudioFadeConfig audioFadeConfig;
+    std::atomic_bool isAudioFadeQueued = false;
+    std::atomic<uint32_t> fadeVolume = kDefaultFadeVolume;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSinkPrivate.h:53]: Audiofade in priv is never set, but we dont set it. So we can remove it and just return the default volume on get failure.

File: source/RialtoGStreamerMSEAudioSinkPrivate.h [hunk 6]
```diff
@@ -38,6 +47,10 @@ struct _RialtoMSEAudioSinkPrivate
     std::atomic_bool isSyncOffQueued = false;
     std::atomic<int32_t> streamSyncMode = kDefaultStreamSyncMode;
     std::atomic_bool isStreamSyncModeQueued = false;
+    std::atomic<const gchar*> audioFade = kDefaultAudioFade;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSinkPrivate.h:50]: No longer need this as you store the struct

File: source/RialtoGStreamerMSEAudioSinkPrivate.h [hunk 7]
```diff
@@ -38,6 +47,10 @@ struct _RialtoMSEAudioSinkPrivate
...
+    AudioFadeConfig audioFadeConfig;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSinkPrivate.h:51]: Probably need a mutex to protect this.

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 8]
```diff
@@ -387,14 +388,11 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
     }
     case PROP_FADE_VOLUME:
     {
-        if (!client)
+        if (!client->getVolume())
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:391]: This isnt how we do it, follow what we do for other properties.

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 9]
```diff
@@ -544,26 +571,29 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
 
         if (sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt) != 3)
         {
-            GST_ERROR_OBJECT(object, "Failed to parse audio fade string");
-            return;
+            GST_WARNING_OBJECT(object, "Failed to parse audio fade string");
         }
 
-        firebolt::rialto::EaseType easeType = static_cast<firebolt::rialto::EaseType>(easeTypeInt);
+        std::lock_guard<std::mutex> lock(priv->audioFadeConfigMutex);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:583]: Reduce the scope of thsi lock, so we dont lock during setVolume

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 10]
```diff
@@ -544,26 +571,29 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
     case PROP_AUDIO_FADE:
     {
         const gchar *audioFadeStr = g_value_get_string(value);
-        double volume;
-        uint32_t duration;
-        int easeTypeInt;
+        double volume = 0.0;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:574]: Reason we dont use the kDefaults here?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 11]
```diff
@@ -544,26 +571,29 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
+        uint32_t duration = 0;
+        int easeTypeInt = convertEaseTypeToInt(firebolt::rialto::EaseType::EASE_LINEAR);
 
         if (sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt) != 3)
         {
-            GST_ERROR_OBJECT(object, "Failed to parse audio fade string");
-            return;
+            GST_WARNING_OBJECT(object, "Failed to parse audio fade string");
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:580]: Print the audioFadeStr in the warning, and print the values that we are going to set.

File: source/RialtoGStreamerMSEBaseSinkPrivate.h [hunk 12]
```diff
@@ -61,6 +61,7 @@ struct _RialtoMSEBaseSinkPrivate
     GstCaps *m_caps = nullptr;
 
     std::atomic<int32_t> m_sourceId;
+    std::atomic<double> m_currentVolume;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEBaseSinkPrivate.h:64]: Where is this used?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 13]
```diff
@@ -76,7 +79,7 @@ static GstStateChangeReturn rialto_mse_audio_sink_change_state(GstElement *eleme
         }
         if (priv->isVolumeQueued)
         {
-            client->setVolume(priv->volume);
+            client->setVolume(priv->targetVolume, priv->volumeDuration, priv->easeType);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:82]: What happens if we have an audio-fade followed by a set volume from the client? Its going to send the previous values of duration and easeType from the audio-fade, so use defaults for these variables.

File: source/RialtoGStreamerMSEAudioSink.cpp (hunk 8 above)
Comment (skyfroxl) [reply source/RialtoGStreamerMSEAudioSink.cpp:400]: You're still not setting getting the volume here and passing it back to client.

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 14]
```diff
@@ -521,6 +577,37 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
+    case PROP_AUDIO_FADE:
+    {
+        const gchar *audioFadeStr = g_value_get_string(value);
+        double volume = kDefaultVolume;
+        uint32_t duration = kDefaultVolumeDuration;
+        int easeTypeInt = convertEaseTypeToInt(kDefaultEaseType);
+
+        if (sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt) != 3)
+        {
+            GST_WARNING_OBJECT(object, "Failed to parse audio fade string: %s. Default values: volume=%lf, duration=%u, easeTypeInt=%d",
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:589]: These might not all be 'Default values' might only parse the first to.
I would say
'Continuing with values:'

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 15]
```diff
@@ -394,14 +394,17 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
     {
         if (!client)
         {
-            g_value_set_uint(value, priv->targetVolume);
+            g_value_set_double(value, priv->targetVolume);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:397]: Is this correct? Is fade-volume property a double?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 16]
```diff
@@ -394,14 +394,17 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
...
             return;
         }
         if (!client->getVolume())
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:400]: This bit isnt required, we dont need to call getVolume twice.

File: source/RialtoGStreamerMSEAudioSink.cpp (hunk 13 above)
Comment (skyfroxl) [reply source/RialtoGStreamerMSEAudioSink.cpp:82]: I dont think you fixed this?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 17]
```diff
@@ -377,6 +390,16 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
         g_value_set_int(value, streamSyncMode);
         break;
     }
+    case PROP_FADE_VOLUME:
+    {
+        if (!client)
+        {
+            g_value_set_uint(value, priv->fadeVolume);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:397]: You never set fadeVolume member, do we need two parameters for volume (fadeVolume/targetVolume)

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 18]
```diff
@@ -377,6 +390,16 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
...
+            return;
+        }
+        g_value_set_uint(value, client->getVolume());
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:400]: get Volume returns a double, you need to converse to uint for fade-volume property. Same above if you use one member.

File: tests/ut/GstreamerMseAudioSinkTests.cpp [hunk 19]
```diff
@@ -434,6 +435,102 @@ TEST_F(GstreamerMseAudioSinkTests, ShouldGetStreamSyncModeProperty)
     gst_object_unref(textContext.m_pipeline);
 }
 
+TEST_F(GstreamerMseAudioSinkTests, ShouldReturnDefaultFadeVolumeValueWhenPipelineIsBelowPausedState)
+{
+    RialtoMSEBaseSink *audioSink = createAudioSink();
+
+    guint fadeVolume{0};
```
Comment (skyfroxl) [inline tests/ut/GstreamerMseAudioSinkTests.cpp:442]: 0 is the default, set it as a value that isnt 0 so we check that it gets set correctly by gstreamer

File: tests/ut/GstreamerMseAudioSinkTests.cpp [hunk 20]
```diff
@@ -434,6 +435,102 @@ TEST_F(GstreamerMseAudioSinkTests, ShouldGetStreamSyncModeProperty)
...
+    gst_object_unref(pipeline);
+}
+
+TEST_F(GstreamerMseAudioSinkTests, ShouldFailWhenParsingInvalidAudioFadeString)
+{
+    TestContext textContext = createPipelineWithAudioSinkAndSetToPaused();
+
+    const gchar *kInvalidFadeConfig{"invalid"};
+
+    EXPECT_CALL(m_mediaPipelineMock, setVolume(_, _, _)).Times(0);
```
Comment (skyfroxl) [inline tests/ut/GstreamerMseAudioSinkTests.cpp:495]: Remove .Times(0) as they are strict mocks

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 21]
```diff
@@ -394,10 +392,10 @@ static void rialto_mse_audio_sink_get_property(GObject *object, guint propId, GV
     {
         if (!client)
         {
-            g_value_set_uint(value, priv->fadeVolume);
+            g_value_set_uint(value, kDefaultFadeVolume);
             return;
         }
-        g_value_set_uint(value, client->getVolume());
+        g_value_set_uint(value, static_cast<unsigned int>(client->getVolume()));
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:398]: Fade volume is 0-100, volume is 0.0-1.0,  need to convert properly

File: source/RialtoGStreamerMSEAudioSinkPrivate.h [hunk 22]
```diff
@@ -21,12 +21,22 @@
...
+{
+    double volume;
+    uint32_t duration;
+    firebolt::rialto::EaseType easeType;
+};
+
 struct _RialtoMSEAudioSinkPrivate
 {
-    std::atomic<double> volume = kDefaultVolume;
+    std::atomic<double> targetVolume = kDefaultVolume;
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSinkPrivate.h:37]: Are these two parameters needed now we have the AudioFadeConfig struct?

File: tests/ut/GstreamerMseAudioSinkTests.cpp [hunk 23]
```diff
@@ -451,10 +451,11 @@ TEST_F(GstreamerMseAudioSinkTests, ShouldGetFadeVolumeProperty)
     TestContext textContext = createPipelineWithAudioSinkAndSetToPaused();
 
     guint fadeVolume{0};
-    constexpr guint kFadeVolume{50};
-    EXPECT_CALL(m_mediaPipelineMock, getVolume(_)).WillOnce(DoAll(SetArgReferee<0>(kFadeVolume), Return(true)));
+    constexpr guint kFadeVolume{5};
```
Comment (skyfroxl) [inline tests/ut/GstreamerMseAudioSinkTests.cpp:454]: Why do we need to const fade volumes?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 24]
```diff
@@ -76,10 +79,19 @@ static GstStateChangeReturn rialto_mse_audio_sink_change_state(GstElement *eleme
...
         }
-
+        if (priv->isAudioFadeQueued)
+        {
+            AudioFadeConfig audioFadeConfig;
+            {
+                std::lock_guard<std::mutex> lock(priv->audioFadeConfigMutex);
+                audioFadeConfig = priv->audioFadeConfig;
+            }
+            client->setVolume(priv->targetVolume, audioFadeConfig.duration, audioFadeConfig.easeType);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:92]: targetVolume isnt set when we set the fade volume,, I think it was correct how it was before audioFadeConfig.volume

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 25]
```diff
@@ -521,6 +571,44 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
         break;
     }
+    case PROP_AUDIO_FADE:
+    {
+        const gchar *audioFadeStr = g_value_get_string(value);
+        double volume = kDefaultVolume;
+        uint32_t duration = kDefaultVolumeDuration;
+        int easeTypeInt = convertEaseTypeToInt(kDefaultEaseType);
+
+        int parsedItems = sscanf(audioFadeStr, "%lf,%u,%d", &volume, &duration, &easeTypeInt);
```
Comment (skyfroxl) [inline source/RialtoGStreamerMSEAudioSink.cpp:581]: Realised that this is incorrect, your parseing a fade volume (0-100) into a double. But dont convert it correctly for setVolume()

**Review by skyfroxl — DISMISSED** [review]

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 26]
```diff
@@ -553,6 +604,45 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
...
+            GST_ERROR_OBJECT(object, "Failed to parse any values from audio fade string: %s.", audioFadeStr);
+            return;
+        }
+        else if (parsedItems == 1 || parsedItems == 2)
+        {
+            GST_WARNING_OBJECT(object, "Partially parsed audio fade string: %s. Continuing with values: fadeVolume=%u, duration=%u, easeTypeInt=%d",
+                               audioFadeStr, fadeVolume, duration, easeTypeInt);
+        }
+
+        double volume = fadeVolume / 100.0;
```
Comment (aczs) [inline source/RialtoGStreamerMSEAudioSink.cpp:627]: Maybe we should make volume clipping here?
If fadeVolume > 100, then clip it to 100 and print warning.

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 27]
```diff
@@ -492,7 +543,7 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
     case PROP_LOW_LATENCY:
     {
         priv->lowLatency = g_value_get_boolean(value);
-        if (!client)
+        if (!client || !basePriv->m_sourceAttached)
```
Comment (aczs) [inline source/RialtoGStreamerMSEAudioSink.cpp:546]: why source needs to be attached, wouldn't caching on the RialtoServer side be enough?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 28]
```diff
@@ -508,7 +559,7 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
     case PROP_SYNC:
     {
         priv->sync = g_value_get_boolean(value);
-        if (!client)
+        if (!client || !basePriv->m_sourceAttached)
```
Comment (aczs) [inline source/RialtoGStreamerMSEAudioSink.cpp:562]: why source needs to be attached, wouldn't caching on the RialtoServer side be enough?

File: source/RialtoGStreamerMSEAudioSink.cpp [hunk 29]
```diff
@@ -524,7 +575,7 @@ static void rialto_mse_audio_sink_set_property(GObject *object, guint propId, co
     case PROP_SYNC_OFF:
     {
         priv->syncOff = g_value_get_boolean(value);
-        if (!client)
+        if (!client || !basePriv->m_sourceAttached)
```
Comment (aczs) [inline source/RialtoGStreamerMSEAudioSink.cpp:578]: why source needs to be attached, wouldn't caching on the RialtoServer side be enough?

**Review by skyfroxl — APPROVED** [review]

**Pull request merged**